server/server.go

```diff
@@ -155,6 +163,247 @@ func New(dependencies *Dependencies, config *Config) (*Server, error) {
 	return server, nil
 }
 
+type ConcurrentSet struct {
+	m map[string]bool
+	sync.RWMutex
+}
+
+func NewConcurrentSet() *ConcurrentSet {
+	return &ConcurrentSet{
+		m: make(map[string]bool),
+	}
+}
+
+func (s *ConcurrentSet) Add(item string) {
+	s.Lock()
+	defer s.Unlock()
+	s.m[item] = true
+}
+
+func (s *ConcurrentSet) Has(item string) bool {
+	s.RLock()
+	defer s.RUnlock()
+	_, ok := s.m[item]
+	return ok
+}
+
+func (s *Server) ListObjects(ctx context.Context, req *openfgapb.ListObjectsRequest) (*openfgapb.ListObjectsResponse, error) {
+	storeID := req.GetStoreId()
+	targetObjectType := req.GetType()
+
+	modelID, err := s.resolveAuthorizationModelID(ctx, storeID, req.GetAuthorizationModelId())
+	if err != nil {
+		return nil, err
+	}
+
+	iter, err := s.datastore.ReadRelationshipTuples(ctx, storage.ReadRelationshipTuplesFilter{
+		StoreID:            storeID,
+		OptionalObjectType: targetObjectType,
+	})
+	if err != nil {
+		s.logger.Error("database query failed", logger.Error(err))
+		return nil, status.Errorf(codes.Internal, "An internal server error has occurred")
+	}
+	defer iter.Stop()
+
+	g := new(errgroup.Group)
+	g.SetLimit(MaximumConcurrentChecks)
+
+	uniqueObjectIdsSet := NewConcurrentSet()
+	uniqueObjectIdsChannel := make(chan string, MaximumConcurrentChecks)
+
+	timeoutCtx := ctx
+	var cancel context.CancelFunc
+	if s.config.ListObjectsDeadline != 0 {
+		timeoutCtx, cancel = context.WithTimeout(ctx, s.config.ListObjectsDeadline)
+		defer cancel()
+	}
+
+	go func() {
+		var objectsFound uint32
+		for {
+			fmt.Println("inside for loop...")
+			time.Sleep(1 * time.Second)
+			if atomic.LoadUint32(&objectsFound) >= s.config.ListObjectsMaxResults {
+				break
+			}
+
+			t, err := iter.Next()
+			if err != nil && err == storage.TupleIteratorDone {
```

Comment:
What if the err is not nil and also not a `TupleIteratorDone`?

Reply:
good point! i need to send the error back to the parent.